### PR TITLE
DMF API supports target attributes update

### PR DESCRIPTION
**Pull request by kaizimmerm**
Long missing feature. DMF now supports attributes the same way as DDI. However, compared to DDI the target is not asked to provide that feedback. Its completely optional.

Pull request contains the feature itself plus DeviceSimlator support for pushing attribute update whenever a new device gets simulated.

Review: @Jonkno @schabdo

### CODE REVIEW

File: examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/SimulationProperties.java [hunk 1]
```diff
@@ -10,31 +10,80 @@
...
+
+    public List<Attribute> getAttributes() {
+        return attributes;
+    }
+
     public List<Autostart> getAutostarts() {
         return this.autostarts;
     }
 
+    public static class Attribute {
```
Comment (Jonkno) [inline examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/SimulationProperties.java]: Missing javadoc

File: examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java [hunk 2]
```diff
@@ -81,7 +90,7 @@ public void finishUpdateProcessWithError(final SimulatedUpdate update, final Lis
      *
```
Comment (Jonkno) [inline examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java]: Also add @param updateResultMessages on line 79

File: examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java [hunk 3]
```diff
@@ -124,25 +133,58 @@ public void sendActionStatusMessage(final String tenant, final ActionStatus acti
...
+     *
+     * @param tenant
+     *            the tenant to create the target
+     * @param targetId
+     *            the ID of the target to create or update
+     */
+    public void updateAttributesOfThing(final String tenant, final String targetId) {
+        sendMessage(spExchange, attributeUpdateMessage(tenant, targetId));
 
+        LOGGER.debug("Send update attributes message and send to update server for Thing \"{}\"", targetId);
```
Comment (Jonkno) [inline examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java]: Better: Create update attributes ...

File: examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java [hunk 4]
```diff
@@ -124,25 +133,58 @@ public void sendActionStatusMessage(final String tenant, final ActionStatus acti
...
      *            the ID of the target to create or update
      */
     public void createOrUpdateThing(final String tenant, final String targetId) {
+        sendMessage(spExchange, thingCreatedMessage(tenant, targetId));
+
+        LOGGER.debug("Created thing created message and send to update server for Thing \"{}\"", targetId);
+    }
+
+    /**
+     * Create new attribute update message and send to udpate server.
```
Comment (Jonkno) [inline examples/hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/SpSenderService.java]: Typo: update

File: hawkbit-dmf-api/src/main/java/org/eclipse/hawkbit/dmf/json/model/AttributeUpdate.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Map of {@link Target} atrributes.
```
Comment (Jonkno) [inline hawkbit-dmf-api/src/main/java/org/eclipse/hawkbit/dmf/json/model/AttributeUpdate.java]: Typo: attributes

File: hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerServiceTest.java [hunk 6]
```diff
@@ -164,6 +165,35 @@ public void createThing() {
...
+        messageProperties.setHeader(MessageHeaderKey.TOPIC, "UPDATE_ATTRIBUTES");
+        final AttributeUpdate attributeUpdate = new AttributeUpdate();
+        attributeUpdate.getAttributes().put("testKey1", "testValue1");
+        attributeUpdate.getAttributes().put("testKey2", "testValue2");
+
+        final Message message = amqpMessageHandlerService.getMessageConverter().toMessage(attributeUpdate,
+                messageProperties);
+
+        final ArgumentCaptor<String> targetIdCaptor = ArgumentCaptor.forClass(String.class);
+        final ArgumentCaptor<Map> attributesCaptor = ArgumentCaptor.forClass(Map.class);
```
Comment (Jonkno) [inline hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerServiceTest.java]: If you use @Captor private ArgumentCaptor<Map<String, String>> attributesCaptor; you can avoid the nested generic problem.

Comment (kaizimmerm) [inline hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerServiceTest.java]: ArgumentCaptor.forClass(Map.class) is actually not able to handle that.

**Pull request merged**